contracts/Safe.sol

```diff
@@ -411,23 +423,52 @@ contract Safe is
         address gasToken,
         address refundReceiver,
         uint256 _nonce
-    ) private view returns (bytes memory) {
-        bytes32 safeTxStructHash = keccak256(
-            abi.encode(
-                SAFE_TX_TYPEHASH,
-                to,
-                value,
-                keccak256(data),
-                operation,
-                safeTxGas,
-                baseGas,
-                gasPrice,
-                gasToken,
-                refundReceiver,
-                _nonce
-            )
-        );
-        return abi.encodePacked(bytes1(0x19), bytes1(0x01), domainSeparator(), safeTxStructHash);
+    ) private view returns (bytes memory result) {
+        bytes32 separatorHash = domainSeparator();
+
+        // The encoding is written in assembly to avoid memory re-allocations.
+        // Straightforward solidity implementation would allocate multiple times, but since the only reason it allocates is because it needs to hash the data,
+        // we can save gas by reusing the same memory block by overwriting the data.
+        /* solhint-disable no-inline-assembly */
+        assembly {
+            let freeMemoryPointer := mload(0x40)
+
+            // Copy the data from calldata to memory and calculate the hash
+            let calldataLength := data.length
+            calldatacopy(freeMemoryPointer, data.offset, calldataLength)
+            let calldataHash := keccak256(freeMemoryPointer, calldataLength)
+
+            // Prepare the SafeTX struct for hashing, overwriting the copied calldata
+            mstore(freeMemoryPointer, SAFE_TX_TYPEHASH)
```

Comment:
naming-nit: `freeMemoryPointer` is a bit long and _slightly_ inaccurate (since this ends up being the result pointer, and the free memory points to after our allocation by the end of the function).

Personally, to be more succinct, I would just call this `ptr`.